### PR TITLE
Allow making context specific requests using the data module

### LINKED ISSUE #11643: Data storage layer should properly handle both context=view and context=edit
For better or for worse, the REST API has an architectural feature called `context`. `context` roughly relates to user permissions, and denotes the context in which the response data will be used:

* `context=view` denotes attributes used for _frontend_ presentation. 
* `context=edit` denotes attributes used for editing the resource.

To provide a specific example:

* `GET /wp-json/wp/v2/posts/1?context=view` returns `post.content.rendered`.
* `GET /wp-json/wp/v2/posts/1?context=edit` also includes `post.content.raw`, but _requires the user to be able to edit the Post_.

Seems simple, right? Not quite.

In the editor context (i.e. Gutenberg), a given WordPress user may have `context=edit` for a Post but only `context=view` for other resources.  For instance, an Editor can edit others' Posts but cannot edit other Users. As such, when the Post Authors UI is presented to the Editor, the REST API request needs to be made with `context=view`. Similarly, Authors and Contributors cannot edit Terms, but they must be able to view them in order to assign Categories and Tags to Posts.

The broader implications of this issue originally surfaced in https://github.com/WordPress/gutenberg/issues/11335#issuecomment-435021606. @youknowriad opened a core Trac ticket https://core.trac.wordpress.org/ticket/45252 to solve this within the REST API, but I'm not optimistic that it's possible. `context` was a topic we discussed at depth years ago — for some past conversation, please read through https://github.com/WP-API/WP-API/issues/702 and linked issues.

It would be more robust if Gutenberg's data layer could handle `context=view` and `context=edit`.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #11643 

We've been struggling with context specific data request for a long time in Gutenberg. This is due to the fact that the data module had always had a single representation of what an entity record could look like.

In this PR, I'm exploring allowing context specific requests. I'm doing so by separating the results from requests of different context, it's like considering these as different entities entirely.

Hopefully, this will allow us to get rid of all the adhoc `apiFetch` usage across the code base.

### CODE REVIEW

**Comment by youknowriad** [comment]
(Don't review this just yet, I have some local changes changing the approach entirely but I'm having trouble pushing them due to internet issues, I'll update here once I'm done)

**Comment by youknowriad** [comment]
Ok I managed to push the changes, this is ready to review.

File: packages/core-data/src/utils/on-sub-key.js [hunk 1]
```diff
@@ -12,7 +17,9 @@ export const onSubKey = ( actionProperty ) => ( reducer ) => (
 ) => {
 	// Retrieve subkey from action. Do not track if undefined; useful for cases
 	// where reducer is scoped by action shape.
-	const key = action[ actionProperty ];
+	const key = isFunction( actionProperty )
```
Comment (ntsekouras) [inline packages/core-data/src/utils/on-sub-key.js]: Why this change? How it can be a function since you added `onSubKey( 'context' )`? In general this part confuses me the `onSubKey`..

File: packages/core-data/src/queried-data/selectors.js [hunk 2]
```diff
@@ -79,7 +88,10 @@ function getQueriedItemsUncached( state, query ) {
 		} else {
 			// If expecting a complete item, validate that completeness, or
 			// otherwise abort.
-			if ( ! state.itemIsComplete[ itemId ] ) {
+			if (
+				! state.itemIsComplete[ context ] ||
+				! state.itemIsComplete[ context ][ itemId ]
```
Comment (ntsekouras) [inline packages/core-data/src/queried-data/selectors.js]: I guess you can use optional chaining here: `! state.itemIsComplete[ context ]?.[ itemId ]`.

Similar with the above `! state.items[ context ].hasOwnProperty( itemId )`.

File: packages/core-data/src/utils/on-sub-key.js (hunk 1 above)
Comment (youknowriad) [reply packages/core-data/src/utils/on-sub-key.js]: This is not needed anymore :) I used it at some point but I can remove it now.

File: packages/core-data/src/selectors.js [hunk 3]
```diff
@@ -147,17 +147,18 @@ export function getEntityRecord( state, kind, name, key, query ) {
...
+		if ( ! queriedState.itemIsComplete[ context ]?.[ key ] ) {
 			return undefined;
 		}
 
-		return queriedState.items[ key ];
+		return queriedState.items[ context ][ key ];
 	}
 
-	const item = queriedState.items[ key ];
+	const item = queriedState.items[ context ][ key ];
```
Comment (ntsekouras) [inline packages/core-data/src/selectors.js]: It can get here without being set for the first time of a specific context value.
```suggestion
	const item = queriedState.items[ context ]?.[ key ];
```

File: packages/core-data/src/queried-data/get-query-parts.js [hunk 4]
```diff
@@ -41,6 +41,7 @@ export function getQueryParts( query ) {
 		perPage: 10,
 		fields: null,
 		include: null,
+		context: 'default',
```
Comment (jsnajdr) [inline packages/core-data/src/queried-data/get-query-parts.js:44]: If the `context` query param is omitted, the default value is `view`.

File: packages/core-data/src/queried-data/reducer.js [hunk 5]
```diff
@@ -71,24 +81,30 @@ export function getMergedItemIds( itemIds, nextItemIds, page, perPage ) {
...
+					}, {} ),
+				},
 			};
+		}
 		case 'REMOVE_ITEMS':
-			const newState = omit( state, action.itemIds );
-			return newState;
+			return mapValues( state, ( contextState ) =>
+				omit( contextState, action.itemIds )
+			);
```
Comment (jsnajdr) [inline packages/core-data/src/queried-data/reducer.js:107]: Does this mean that query results for different contexts are stored separately, because the item data are actually different, while `removeItems` always removes the cached data for all contexts without discrimination?

**Comment by youknowriad** [comment]
I'm adding this to the backport list to be able to use it to fix the permissions checks for the site blocks.

**Review by ntsekouras — APPROVED** [review]
That's a great addition and you did it amazingly fast after discussing it yesterday 🚀 !

Great work Riad!!

**Review by jsnajdr — COMMENTED** [review]
The `packages/core-data/src/reducer` reducer also handles the `RECEIVE_ITEMS` action in the `edits` part. Should it also check the `action.context` field and ignore responses that don't have the `edit` context?

The `receiveEntityRecords` and `removeItems` have an `invalidateCache` flag. The `getEntityRecords.shouldInvalidate` handler that looks at this flag should correctly and carefully choose the queries to invalidate: only the context-specific one for `RECEIVE_ITEMS`, and all contexts for `REMOVE_ITEMS`.

File: packages/core-data/src/queried-data/get-query-parts.js (hunk 4 above)
Comment (youknowriad) [reply packages/core-data/src/queried-data/get-query-parts.js:44]: no, we do enforce "edit" context in the API requests, even if the context is "view" by default for the server, that's why I opted on a new "default"/"empty" context in the frontend. Backward compatibility.

File: packages/core-data/src/queried-data/reducer.js (hunk 5 above)
Comment (youknowriad) [reply packages/core-data/src/queried-data/reducer.js:107]: yes

**Comment by youknowriad** [comment]
> The receiveEntityRecords and removeItems have an invalidateCache flag. The getEntityRecords.shouldInvalidate handler that looks at this flag should correctly and carefully choose the queries to invalidate: only the context-specific one for RECEIVE_ITEMS, and all contexts for REMOVE_ITEMS.

I think creating a record in a one context, should invalidate queries in all contexts

File: packages/core-data/src/queried-data/get-query-parts.js (hunk 4 above)
Comment (jsnajdr) [reply packages/core-data/src/queried-data/get-query-parts.js:44]: Does that mean that calling `getQueriedItems` without `context` will issue an API request with `?context=edit` query param and will store the result under the `state[ 'default' ]` key?

And that `getQueriedItems( state, { context: 'edit' } )` will issue a request with the same `?context=edit` query param but this time will store the result as `state[ 'edit' ]`?

Comment (youknowriad) [reply packages/core-data/src/queried-data/get-query-parts.js:44]: > Does that mean that calling getQueriedItems without context will issue an API request with ?context=edit query param and will store the result under the state[ 'default' ] key?

In most entities yes, but entities can define another set of default REST API parameters. 

I agree it's not ideal that `edit` is different than `default` in this cases, but solving it properly would have required a small change in the entities API (explicitly define a default context instead of relying on a more global baseURLParams). I think it's something we can try but I thought it was not important for v1 of this.

Comment (jsnajdr) [reply packages/core-data/src/queried-data/get-query-parts.js:44]: I agree that it's not terribly important. I mostly wanted to confirm that this little flaw is really there, and that it's not just my misunderstanding.
